crates/ruff_python_formatter/src/expression/string.rs

```diff
@@ -1080,74 +1144,537 @@ fn format_docstring(normalized: &NormalizedString, f: &mut PyFormatter) -> Forma
+            let indent_len = indentation_length(trim_end) - self.stripped_indentation_length;
+            let in_docstring_indent = " ".repeat(usize::from(indent_len)) + trim_end.trim_start();
+            text(&in_docstring_indent, Some(line.offset)).fmt(self.f)?;
         } else {
-            empty_line().fmt(f)
+            // Take the string with the trailing whitespace removed, then also
+            // skip the leading whitespace.
+            let trimmed_line_range = TextRange::at(line.offset, trim_end.text_len())
+                .add_start(self.stripped_indentation_length);
+            if self.already_normalized {
+                source_text_slice(trimmed_line_range).fmt(self.f)?;
+            } else {
+                // All indents are ascii spaces, so the slicing is correct.
+                text(
+                    &trim_end[usize::from(self.stripped_indentation_length)..],
+                    Some(trimmed_line_range.start()),
+                )
+                .fmt(self.f)?;
+            }
+        }
+
+        // We handled the case that the closing quotes are on their own line
+        // above (the last line is empty except for whitespace). If they are on
+        // the same line as content, we don't insert a line break.
+        if !line.is_last {
+            hard_line_break().fmt(self.f)?;
+        }
+
+        Ok(())
+    }
+
+    /// Given a sequence of lines from a code snippet, format them and return
+    /// the formatted code as a sequence of owned docstring lines.
+    ///
+    /// This routine generally only returns an error when the recursive call
+    /// to the formatter itself returns a `FormatError`. In all other cases
+    /// (for example, if the code snippet is invalid Python or even if the
+    /// resulting reformatted code snippet is invalid Python), then `Ok(None)`
+    /// is returned. In this case, callers should assume that a reformatted
+    /// code snippet is unavailable and bail out of trying to format it.
+    ///
+    /// Currently, when the above cases happen and `Ok(None)` is returned, the
+    /// routine is silent about it. So from the user's perspective, this will
+    /// fail silently. Ideally, this would at least emit a warning message,
+    /// but at time of writing, it wasn't clear to me how to best do that.
+    fn format(
+        &mut self,
+        code: &[CodeExampleLine<'src>],
+    ) -> FormatResult<Option<Vec<DocstringLine<'static>>>> {
+        use ruff_python_parser::AsMode;
+
+        let offset = code
+            .get(0)
+            .expect("code blob must be non-empty")
+            .original
+            .offset;
+        let last_line_is_last = code
+            .last()
+            .expect("code blob must be non-empty")
+            .original
+            .is_last;
+        let codeblob = code
+            .iter()
+            .map(|line| &*line.code)
+            .collect::<Vec<&str>>()
+            .join("\n");
```

Comment:
Hmm, we may need to preserve line endings here, right? What if the code uses non-LF line endings, or the formatter has a line ending set?

Reply:
I don't believe so, no. I could use `\r\n` here and things would still turn out okay. I did it this way because immediately after this, we reformat the code snippet with the user's settings, which should include their line ending preferences.

With that said... I tried to write a test for this and failed. I _suspect_ something is mangling or normalizing line endings in our test infrastructure. If I run the formatter directly, then line endings are preserved:

```
$ rg '\x0D' /tmp/test.py
$ cat /tmp/ruff.toml
[format]
docstring-code = true
$ cargo run -qp ruff_cli format --config /tmp/ruff.toml - < /tmp/test.py | rg '\x0D'
$ cat /tmp/ruff.toml
[format]
docstring-code = true
line-ending = 'cr-lf'
$ cargo run -qp ruff_cli format --config /tmp/ruff.toml - < /tmp/test.py | rg '\x0D'
def doctest_line_ending():
    """
    Do cool stuff.
    >>> def foo(x):
    ...     print(x)
    ...
    ...     print(x)
    """
    pass
```

I'll take a look and see if I can figure out what's swallowing line endings. When I add a test with `line-ending = "CarriageReturnLineFeed"`, the resulting snapshot does not contain CRLF line endings, but LF only.

Reply:
Take a look at the `.gitattributes` in the repo root. I believe we normalize to LF except in explicit cases.